### DISCUSSION TITLE
How could extract username from auth_User like 'domain\username'

### DISCUSSION THREAD

**Discussion by efcjunior** [DirectoryTree/LdapRecord#486] [General]
The auth_user or remote_user variables returns something like 'domain\username'. That's way I can't use the WindowsAuthenticate.  Is there anyway to extract username from these variables?

**Comment by stevebauman** [comment]
```php
[$domain, $username] = explode('\\', $_SERVER['auth_user']);
```

Does this answer your question?

**Comment by efcjunior** [comment]
> ```
> [$domain, $username] = explode('\\', $_SERVER['auth_user']);
> ```
> 
> Does this answer your question?

Yep! But which class should I put it ? 

Thank you.

**Comment by stevebauman** [comment]
I'd say wherever you need it... I'm not exactly sure what you're attempting to implement, I'd need more information.

**Reply by efcjunior** [reply]
> I'd say wherever you need it... I'm not exactly sure what you're attempting to implement, I'd need more information.

I'm trying to configure SSO like this:

protected $middlewareGroups = [
    'web' => [
        // ...
        \LdapRecord\Laravel\Middleware\WindowsAuthenticate::class,
    ],
];

and: 

Route::get('/', LoginController::class)->middleware('web');

But, when I call var_dump(Auth::user()); from LoginController it returns null. 

I've already tested AUTH DATABASE and it worked perfectly, I'm a bit lost how to configure SSO. That's why I think the problem could be the AUTH_USER format.

**Reply by stevebauman** [reply]
The format you've shown is the standard SSO format. The included middleware will parse it for you 👍 

https://ldaprecord.com/docs/laravel/v2/auth/sso/setup#sso-domain-verification

**Comment by efcjunior** [comment]
I think the WindowsAuthenticate is not loading because when I do this I got error Class 'App\Providers\WindowsAuthenticate' not found 


// app/Providers/AuthServiceProvider.php

/**
 * Register any authentication / authorization services.
 *
 * @return void
 */
public function boot()
{
    $this->registerPolicies();

    WindowsAuthenticate::serverKey('PHP_AUTH_USER');
}

**Reply by stevebauman** [reply]
You have to import the class:

```php
use LdapRecord\Laravel\Middleware\WindowsAuthenticate;

// ...

public function boot()
{
    $this->registerPolicies();

    WindowsAuthenticate::serverKey('PHP_AUTH_USER');
}
```

**Reply by efcjunior** [reply]
> You have to import the class:
> 
> ```
> use LdapRecord\Laravel\Middleware\WindowsAuthenticate;
> 
> // ...
> 
> public function boot()
> {
>     $this->registerPolicies();
> 
>     WindowsAuthenticate::serverKey('PHP_AUTH_USER');
> }
> ```

Oh! Sorry. That's true. What's the right way to verify if the user is authenticated on Controller ? When I assign the WindowsAuthenticate middleware to a route and it fails, the request reach out the controller restricted.

**Reply by stevebauman** [reply]
Use the `auth` middleware included in Laravel, either on your route, or your controller constructor:

```php
Route::middleware('auth')->get('dashboard', [DashboardController::class, 'index']);
```

```php
class DashboardController
{
    public function __construct()
    {
        $this->middleware('auth');
    }
}
```

**Reply by efcjunior** [reply]
> Use the `auth` middleware included in Laravel, either on your route, or your controller constructor:
> 
> ```
> Route::middleware('auth')->get('dashboard', [DashboardController::class, 'index']);
> ```
> 
> ```
> class DashboardController
> {
>     public function __construct()
>     {
>         $this->middleware('auth');
>     }
> }
> ```

Hi @stevebauman 

It worked perfectly! Now my application has configured DATABASE AUTH + SSO together. 

Could you help me to figure out these situations:

1) Is there anyway to add auth_user variable in development environment using php artisan serve ?
